drivers/sensor/bmi160/bmi160_v2.c

```diff
@@ -0,0 +1,364 @@
+
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(781)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(1563)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(3125)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(6250)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(12500)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(25000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(50000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(100000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(200000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(400000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(800000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(1600000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GYROSCOPE, UINT32_C(3200000)),
+
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(781)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(1563)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(3125)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(6250)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(12500)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(25000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(50000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(100000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(200000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(400000)),
+		BUILD_SAMPLE_RATE_INFO(SENSOR_TYPE_GEOMAGNETIC_FIELD, UINT32_C(800000)),
+	};
+
+	*sample_rates = bmi160_sample_rates;
+	*count = ARRAY_SIZE(bmi160_sample_rates);
+	return 0;
+}
+
+static int bmi160_set_watermark(const struct device *sensor, uint8_t watermark_percent,
+				bool round_up)
+{
+	const struct bmi160_cfg *cfg = sensor->config;
+	uint8_t watermark_reg_value = 0;
+	uint8_t int_status_1;
+
+	watermark_percent = CLAMP(watermark_percent, 0, 100);
+
+	if (watermark_percent != 0) {
+		/* BMI160 watermark is 4 bytes per 1 value in the register.
+		 * First convert the watermark_percent to an actual value relative to the FIFO size
+		 * (1024):
+		 *     1024 * watermark_percent / 100
+		 * Then divide by 4 to get:
+		 *     (1024 * watermark_percent) / 400
+		 */
+		watermark_reg_value = (1024 * watermark_percent) / 400;
+		if (round_up && ((1024 * watermark_percent) / 400) != 0) {
+			watermark_reg_value++;
+		}
+		if (cfg->bus_io->write(sensor, BMI160_REG_FIFO_CONFIG0, &watermark_reg_value, 1) !=
+		    0) {
+			return -EIO;
+		}
+	} else {
+		if (cfg->bus_io->write(sensor, BMI160_REG_FIFO_CONFIG0, &watermark_reg_value, 1) !=
+		    0) {
+			return -EIO;
+		}
+	}
+
+	/* Read the current watermark config */
+	if (cfg->bus_io->read(sensor, BMI160_REG_INT_STATUS1, &int_status_1, 1) != 0) {
+		return -EIO;
```

Comment:
What motivates the use of `cfg->bus_io->read` vs `bmi160_read` and friends in this file?